File: 09_DataStructure_Algorithm/0513_day48/03_가장 먼 노드/가장 먼 노드_bfs.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {        
        List<List<Integer>> graph = new ArrayList<>(n);
        
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        
        for(int[] e : edge) {
            graph.get(e[0] - 1).add(e[1] - 1);
            graph.get(e[1] - 1).add(e[0] - 1);
        }
        
        int[] distance = new int[n];
        boolean[] visited = new boolean[n];
        Queue<Integer> queue = new ArrayDeque<>();
        
        int start = 0;
        queue.offer(start);
        visited[start] = true;
        distance[start] = 0;
        
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            
            for (int nxt : graph.get(cur)) {
                if (!visited[nxt]) {
                    visited[nxt] = true;
                    distance[nxt] = distance[cur] + 1;
                    queue.offer(nxt);
                }
            }
        }
        
        int max = distance[start];
        for (int i = start; i < n; i++) {
            if (max < distance[i]) max = distance[i];
        }
        
        int answer = 0;
        for (int d : distance) {
            if (d == max) answer++;
        }
        
        return answer;
    }
}

File: 09_DataStructure_Algorithm/0513_day48/03_가장 먼 노드/가장 먼 노드_bfs.py
from collections import deque

def solution(n, edge):
    graph = [[] for _ in range(n)]
    
    for e in edge:
        e1, e2 = e
        graph[e1 - 1].append(e2 - 1)
        graph[e2 - 1].append(e1 - 1)
        
    visited = [-1] * n
    
    start = 0
    queue = deque([start])
    visited[start] = 0
    
    while queue:
        cur = queue.popleft()
        
        for nxt in graph[cur]:
            if visited[nxt] == -1:
                visited[nxt] = visited[cur] + 1
                queue.append(nxt)
    
    return visited.count(max(visited))